jdaviz/configs/cubeviz/plugins/tools.py

```diff
@@ -42,3 +47,146 @@ def on_mouse_event(self, data):
+                if len(data.shape) != 3:
+                    continue
+                cube_in_viewer = True
+
+                cube = data.get_object(cls=Spectrum1D, statistic=None)
+                if x > cube.shape[0] or y > cube.shape[1] or x < 0 or y < 0:
+                    continue
+
+                spec = Spectrum1D(flux=cube.flux[x][y], spectral_axis=cube.spectral_axis,
+                                  meta=cube.meta)
+                # Add meta data for reference data and pixel
+                spec.meta["reference_data"] = data.label
+                spec.meta["created_from_pixel"] = f"({x}, {y})"
+                spec.meta["Plugin"] = "Spectrum per spaxel"
+
+                label = f"{data.label}_at_pixel"
+
+                # Remove data from viewer, re-add data to app, and then add data
+                # back to spectrum viewer
+                self.viewer.jdaviz_app.remove_data_from_viewer("spectrum-viewer", label)
+                dc[label] = spec
+                self.viewer.jdaviz_app.add_data_to_viewer("spectrum-viewer",
+                                                          label,
+                                                          clear_other_data=False)
+                self.created_data_labels.append(label)
+
+                # Set color of spectrum to not conflict with subset colors
+                for layer in self.spectrum_viewer.state.layers:
+                    if layer.layer.label == label:
+                        layer.color = self.colors[0]
+
+                # Link newly created spectrum to reference data
+                ref_cube = self.spectrum_viewer.state.reference_data
+                new_spec = dc[label]
+                self.link_data(ref_cube, new_spec)
+
+            # Flux viewer is empty or contains only 2D images
+            if not cube_in_viewer:
+                self.remove_highlight_on_pixel()
+                msg = SnackbarMessage("No cube in viewer, cannot create spectrum from pixel",
+                                      sender=self, color="warning")
+                self.viewer.session.hub.broadcast(msg)
+                return
+
+    def link_data(self, data1, data2):
+        # If spectrum is also reference data
+        if data1.label == data2.label:
+            return
+
+        world1 = data1.world_component_ids
+        world2 = data2.world_component_ids
+        # Reference data is cube
+        if len(world1) == 3 and len(world2) == 1:
+            links = [LinkSame(world1[2], world2[0])]
+        # Reference data is 1D spectrum
+        elif len(world1) == 1 and len(world2) == 1:
+            links = [LinkSame(world1[0], world2[0])]
+        else:
+            return
+
+        self.viewer.session.data_collection.add_link(links)
+
+    def highlight_pixel(self, x, y):
+        # Creates an X that shows the pixel at coordinate (x, y)
+        x_coords = [[x - 0.5, x + 0.5], [x - 0.5, x + 0.5]]
+        y_coords = [[y - 0.5, y + 0.5], [y + 0.5, y - 0.5]]
```

Comment:
I see repeated math. Should we only calculate each unique equation once and reuse?